core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java

```diff
@@ -389,22 +377,63 @@ protected double getFirstHighEstimate(Envelope envelope) {
     }
 
     @Override
-    public Envelope apply(Envelope base) {
-        var baseTime = getMarginTime(base);
-        var distance = sectionEnd - sectionBegin;
-        var allowanceAddedTime = allowanceValue.getAllowanceTime(baseTime, distance);
-        // if no time is added, just return the base envelope without performing binary search
-        if (allowanceAddedTime == 0.0)
-            return base;
-        var targetTime = baseTime + allowanceAddedTime;
+    public Envelope apply(Envelope base, double[] stopPositions) {
+        var totalBuilder = new EnvelopeBuilder();
+        var partsBefore = base.slice(Double.NEGATIVE_INFINITY, sectionBegin);
+        for (var part : partsBefore)
+            totalBuilder.addPart(part);
+        // Loop to decompose the envelope between stops
+        var cursor = sectionBegin;
+        for (double stopPosition : stopPositions) {
+            if (stopPosition <= sectionBegin)
+                continue;
+            if (stopPosition >= sectionEnd)
+                break;
+
+            var newBase = Envelope.make(base.slice(cursor, stopPosition));
+            var physicalLimits = generatePhysicalLimits(newBase);
+            var search = initializeBinarySearch(newBase);
+            // loop to run binarySearch for this ROI
+            var resultEnvelope = binarySearchLoop(search, newBase, physicalLimits);
+            // add all resulting envelope parts
+            for (int i = 0; i < resultEnvelope.size(); i++)
+                totalBuilder.addPart(resultEnvelope.get(i));
+
+            cursor = stopPosition;
+        }
+        // last
+        var newBase = Envelope.make(base.slice(cursor, sectionEnd));
+        var physicalLimits = generatePhysicalLimits(newBase);
+        var search = initializeBinarySearch(newBase);
+        // loop to run binarySearch for this ROI
+        var resultEnvelope = binarySearchLoop(search, newBase, physicalLimits);
+        for (int i = 0; i < resultEnvelope.size(); i++)
+            totalBuilder.addPart(resultEnvelope.get(i));
```

Comment:
This code is duplicated, we could move it into a new method, something like 

```java
private void applyMarecoToSection(Envelope base, EnvelopeBuilder totalBuilder, double begin, double end)
```

Reply:
Fixed